packages/core/src/render3/util/attrs_utils_with_hydration_support.ts

```diff
@@ -0,0 +1,108 @@
+ * Values are implemented as a list in case they need to be expanded in the future to accommodate
+ * more cases.
+ */
+const hydrationProtectedElementToAttributeMap = new Map<string, string>([
+  // All these elements and their attributes force the browser to reload resources when they're set
+  // again with the same value. For example, consider an `<object>` element with its `data`
+  // attribute set to `/assets/some-file.pdf`. When the browser retrieves an HTML document from the
+  // server and finishes parsing it (after the document state is set to `complete`), it loads
+  // external resources such as images, videos, audios, etc. This includes loading
+  // `/assets/some-file.pdf`. Subsequently, when Angular begins its hydration process, it attempts
+  // to call `setAttribute` on the `<object>` element again with `setAttribute('data',
+  // '/assets/some-file.pdf')`. This action forces the browser to reload the same resources, even
+  // though they have already been loaded previously.
+  ['iframe', 'src'],
+  ['embed', 'src'],
+  ['object', 'data'],
+]);
+
+export function getHydrationProtectedAttribute(tagName: string): string | undefined {
+  return hydrationProtectedElementToAttributeMap.get(
+    // Convert to lowercase so we cover both cases when the tag name is `iframe` or `IFRAME`.
+    tagName.toLowerCase(),
+  );
+}
+
+let _setAttributeWithHydrationSupport: typeof _setAttributeWithHydrationSupportImpl = (
+  isFirstPass: boolean,
+  hydrationInfo: DehydratedView | null,
+  renderer: Renderer,
+  nodeIndex: number,
+  element: RElement,
+  attributeName: string,
+  attributeValue: string,
+  namespace?: string | null,
+) => {
+  renderer.setAttribute(element, attributeName, attributeValue, namespace);
+};
+
+function _setAttributeWithHydrationSupportImpl(
+  isFirstPass: boolean,
+  hydrationInfo: DehydratedView | null,
+  renderer: Renderer,
+  nodeIndex: number,
+  element: RElement,
+  attributeName: string,
+  attributeValue: string,
+  namespace?: string | null,
+) {
+  // `isFirstPass` is required to check whether we are in the process of creating/updating a view
+  // for the first time. We will not perform these checks during subsequent change detection cycles.
+  if (
+    isFirstPass &&
+    getHydrationProtectedAttribute(element.tagName) &&
+    hydrationInfo?.protectedAttributes?.has(nodeIndex)
+  ) {
+    // If the index of the currently hydrating element is found in the `protectedAttributes` map, it
+    // indicates that this element has hydration-protected attributes (for example, if the `element`
+    // is an `iframe` with its `src` attribute set).
+    const protectedAttributeValue = hydrationInfo.protectedAttributes.get(nodeIndex)!;
+    if (attributeValue === protectedAttributeValue) {
+      // We do not want to reset the value since it may negatively affect performance and UX. For
+      // example, in the `<iframe src="...">` case, this would cause a reload within the iframe.
+      return;
```

Comment:
@arturovt I was thinking if we can protect an attribute value for the first "write" operation only and remove a corresponding key from the map afterwards. This should remove the need to pass the `isFirstPass` flag as an argument. What do you think?
```suggestion
    const { protectedAttributes } = hydrationInfo;
    const protectedAttributeValue = protectedAttributes.get(nodeIndex)!;
    if (attributeValue === protectedAttributeValue) {
      // An initial value is protected for the first `setAttribute` call, further protection would
      // happen as a part of the change detection mechanisms. Remove the node index entry
      // from the map to indicate that no special logic are required for this element and a given
      // attribute.
      protectedAttributes.delete(nodeIndex);
      
      // We do not want to reset the value since it may negatively affect performance and UX. For
      // example, in the `<iframe src="...">` case, this would cause a reload within the iframe.
      return;
```

Reply:
Makes sense, I've updated this code path.

Reply:
@arturovt I still see the `isFirstPass` being used in the code of this PR. Please let me know once the PR is updated, I can take another look.